theano/scan_module/scan_op.py

```diff
@@ -1100,9 +1159,19 @@ def execute(self, node, args, outs):
                 output_storage[pdx].storage[0] = None
 
             # 4.5. Keep a reference to the variables currently in the
-            # output_storage to be able to compare them with the actual
-            # outputs of the inner function after its execution
-            old_output_storage = [o.storage[0] for o in output_storage]
+            # output_storage, and their data, to be able to compare them with
+            # the actual outputs of the inner function after its execution
+            for idx in xrange(len(output_storage)):
+
+                var = output_storage[idx].storage[0]
+                old_output_storage[idx] = var
+
+                if hasattr(var, 'gpudata'):
+                    old_output_data[idx] = var.gpudata
+                elif hasattr(var, 'data'):
+                    old_output_data[idx] = var.data
```

Comment:
What does it mean to have `data` or `gpudata`? Are you matching constants? Shared variables? Can you add a comment?